### DISCUSSION TITLE
PathLayer total path uv vs segment uv in fragment shader

### DISCUSSION THREAD

**Discussion by GarrettCannon** [visgl/deck.gl#5911] [Q&A] [answered]
Hi, 
I'm not very experienced with shaders and etc, but how would I go about getting the `uv` value in the fragment shader relative to the whole path rather than relative to the current segment? 
Thanks!

**Comment by Pessimistress** [comment, marked as answer]
This is not possible in shader alone; part of it has to be calculated on the CPU. You can take a look at [PathStyleExtension](https://github.com/visgl/deck.gl/blob/master/modules/extensions/src/path-style/path-style.js)'s `instanceDashOffset` attribute.

**Reply by GarrettCannon** [reply]
Thanks for the quick reply.
~~Does that transform run before or after the update function?~~
That solved my issue, my next question was because I misunderstood what the extension was doing, but I get it now. Thanks again, you've been a huge help!

**Reply by GarrettCannon** [reply]
@Pessimistress I really appreciate the help you've been, so I hate to ask for more. But I am absolutely stumped on this. As seen in the gif below, my dash animation isn't syncing up properly and I am at pretty much a loss as to why.

![non-sync](https://user-images.githubusercontent.com/44881574/123502494-dff80800-d61a-11eb-954a-88f97a699326.gif)

The sync difference seems different each time I load the page and at different zoom levels.
This is using pixel `widthUnits`.

Code:
```
class ShippingLayer extends PathLayer<
  ShipRoutesGeo,
  ShippingLayerProps<ShipRoutesGeo>
> {
  initializeState(params: any) {
    super.initializeState(params);
    this.getAttributeManager().addInstanced({
      instanceOutbound: {
        accessor: "getOutbound",
        defaultValue: 0,
        size: 1,
      },
      instanceOffsetDistance: {
        size: 1,
        accessor: "getPath",
        transform: this.getDashOffsets,
      },
    });
  }

  draw(params: any) {
    params.uniforms.time = performance.now() / 1000;
    super.draw(params);
  }

  getShaders() {
    const shaders = super.getShaders();
    shaders.inject = {
      "vs:#decl": `
        attribute float instanceOutbound;
        attribute float instanceOffsetDistance;
        varying float lineOffset;
        varying float outbound;
      `,
      "vs:#main-end": `
        lineOffset = instanceOffsetDistance;
        outbound = instanceOutbound;
      `,
      "fs:#decl": `
        uniform float time;
        varying float lineOffset;
        varying float outbound;
        vec3 lineColor = vec3(0.368627, 0.862745, 1.0);
        vec3 dashColor = vec3(0.007843, 0.196078, 0.411765);
      `,
      "fs:#main-end": `
        float dash = lineOffset + vPathPosition.y;
        dash *= 0.01;
        dash -= time * 0.5;
        dash = mod(dash, 1.0);
        dash = floor(dash + 0.1);
        gl_FragColor.xyz = mix(lineColor, dashColor, dash);
        gl_FragColor.a = 1.0;
      `
    }
    return shaders;
  }

  getDashOffsets(path: Path) {
    const result = [0];
    const positionSize = this.props.positionFormat === "XY" ? 2 : 3;
    const isNested = Array.isArray(path[0]);
    const geometrySize = isNested ? path.length : path.length / positionSize;

    let p;
    let prevP;
    for (let i = 0; i < geometrySize - 1; i++) {
      p = isNested
        ? path[i]
        : path.slice(i * positionSize, i * positionSize + positionSize);
      p = this.projectPosition(p);

      if (i > 0) {
        result[i] = result[i - 1] + gl.vec3.dist(prevP, p);
      }

      prevP = p;
    }
    return result;
  }
}
```

Again, I greatly appreciate any help you can give, even if it's just pointing me in the right direction resource-wise.
Thanks!

**Reply by GarrettCannon** [reply]
So I actually figured this out on my own, I was missing a line from the PathStyleLayer Extension. But I'll leave my question up for any future people that might have the same issue.
The solution was to divide the offset by `width.x` in the vertex shader:
`lineOffset = instanceOffsetDistance / width.x;`

![line-sync](https://user-images.githubusercontent.com/44881574/123522053-dd33fc00-d688-11eb-9934-40a1333c0118.gif)

**Comment by ibgreen** [comment]
@GarrettCannon This looks really nice! Would you consider contributing this layer? 

It could be a new layer, or perhaps it can be packaged up as a PathLayer extension?

**Reply by GarrettCannon** [reply]
I would definitely consider it. It may have to be a new layer with some PathLayer features removed to make room for more attributes, I think PathLayer sits at 14 attributes by default. It won't be something I can do right away, but maybe in a couple weeks I can take another look.
